File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(number) do
    before_number = number - 1
    after_number = number - 2

    """
    #{describe(before_number)}
    #{take_action(before_number)}, #{bottles_on_the_wall(after_number)}.
    """
  end

  defp describe(number) do
    "#{String.capitalize(bottles_on_the_wall(number))}, #{bottles(number)}."
  end

  defp bottles_on_the_wall(number) do
    "#{bottles(number)} on the wall"
  end

  defp take_action(0) do
    "Go to the store and buy some more"
  end
  defp take_action(number) do
    "Take #{grep_from(number)} down and pass it around"
  end

  defp bottles(number), do: "#{remained(number)} #{unit(number)} of beer"

  defp remained(-1), do: 99 
  defp remained(0), do: "no more"
  defp remained(number), do: number

  defp unit(1), do: "bottle"
  defp unit(_), do: "bottles"

  defp grep_from(1), do: "it"
  defp grep_from(_), do: "one"

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range \\ 100..1) do
    range
    |> Enum.map(&(BeerSong.verse(&1)))
    |> Enum.join("\n")
  end
end

File: python/beer-song/beer_song.py
def verse(n):
    if n == 0:
        return \
            'No more bottles of beer on the wall, no more bottles of beer.\n' \
            'Go to the store and buy some more, ' \
            '99 bottles of beer on the wall.\n'
    elif n == 1:
        return \
            '1 bottle of beer on the wall, 1 bottle of beer.\n' \
            'Take it down and pass it around, ' \
            'no more bottles of beer on the wall.\n'
    elif n == 2:
        return \
            '2 bottles of beer on the wall, 2 bottles of beer.\n' \
            'Take one down and pass it around, 1 bottle of beer on the wall.\n'
    else:
        return \
            f'{n} bottles of beer on the wall, {n} bottles of beer.\n' \
            f'Take one down and pass it around, ' \
            f'{n-1} bottles of beer on the wall.\n'


def song(start, end = 0):
    return '\n'.join(verse(i) for i in range(start, end - 1, -1)) + '\n'
